CriteoPublisherSdk/Sources/Util/CR_URLResolver.m

```diff
@@ -0,0 +1,121 @@
+//
+//  CR_URLResolver.m
+//  CriteoPublisherSdk
+//
+//  Copyright © 2018-2020 Criteo. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "CR_URLResolver.h"
+#import "Logging.h"
+#import "CR_DeviceInfo.h"
+#import "CR_URLRequest.h"
+
+@interface CR_URLResolution ()
+@property(nonatomic, strong) NSURL *URL;
++ (instancetype)resolutionWithType:(CR_URLResolutionType)type;
++ (instancetype)resolutionError:(NSError *)error;
+@end
+
+@interface CR_URLResolver ()
+@end
+
+@interface CR_URLResolver () <NSURLSessionDataDelegate>
+@property(nonatomic, strong, readonly) NSURLSession *urlSession;
+@property(nonatomic, copy) CR_URLResolutionHandler resolution;
+@end
+
+@implementation CR_URLResolver
+
+#pragma mark - Lifecycle
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _urlSession =
+        [NSURLSession sessionWithConfiguration:NSURLSessionConfiguration.defaultSessionConfiguration
+                                      delegate:self
+                                 delegateQueue:nil];
+  }
+  return self;
+}
+
+#pragma mark - Resolve
+
++ (void)resolveURL:(NSURL *)url
+        deviceInfo:(CR_DeviceInfo *)deviceInfo
+        resolution:(CR_URLResolutionHandler)resolution {
+  CR_URLResolver *resolver = [[CR_URLResolver alloc] init];
+  [resolver resolverURL:url deviceInfo:deviceInfo resolution:resolution];
+}
+
+- (void)resolverURL:(NSURL *)url
+         deviceInfo:(CR_DeviceInfo *)deviceInfo
+         resolution:(CR_URLResolutionHandler)resolution {
+  self.resolution = resolution;
+  CR_URLRequest *request = [CR_URLRequest requestWithURL:url deviceInfo:deviceInfo];
+  NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:request];
+  [task resume];
```

Comment:
You can add a completion handler in the task and use it with the threadManager. So the threadManager can be used to wait on such task.

See https://github.com/criteo/ios-publisher-sdk/blob/main/CriteoPublisherSdk/Sources/Network/CR_NetworkManager.m#L85

Reply:
The completion handler is only called when the request is fully loaded, which includes redirections. The main purpose of the resolver is the [`willPerformHTTPRedirection:`](https://developer.apple.com/documentation/foundation/nsurlsessiontaskdelegate/1411626-urlsession) delegate message _before_ the url is loaded, I hardly see how it can be done the way you suggest, but maybe you have another idea?

Reply:
Not an issue, the idea would be to have a way to wait for the end of the processus (so when redirections are over) so that tests can observe an idle state (simpler to assert than during a dynamic state).

Although, this is only useful if we have a functional scenario such as:
- start a stuff
  - stuff trigger redirection
  - once redirection is finished, post a foo
- wait for idle state
- observe that foo is posted

I don't know if you plan to write such a test, if not then ok to skip this.

Reply:
I plan to mock the resolver for tests that will depend on it.

To be honest, generally I prefer to use the "wait for `XCTTestExpectation`" approach that is the "standard" way to test asynchronous code on iOS platform. Most of the case I find it simpler and less intrusive than the threadmanager. Thread manager makes more sense when queuing / concurrency is involved.